### PR TITLE
[tl_detector] Update code, add manual traffic light and also fix bug …

### PR DESCRIPTION
**Pull request by anonymint**
Couple of things here
* [waypoint_updater] fix bug car didn't stop due to traffic data is `data.data`
* [tl_detector] add rate to control freq of publisher and that would be used to control classifier for future, logic is the same no behavior changes
* [tl_detector] Add manual light state for testing purpose, I added that in just for manual test which won't be used when using classifier as mentioned in lecture just convenience way to test traffic light 

I tested locally and car did stop at traffic light

### CODE REVIEW

File: ros/src/tl_detector/tl_detector.py [hunk 1]
```diff
@@ -50,7 +51,39 @@ def __init__(self):
...
+        rate = rospy.Rate(freq)
+        while not rospy.is_shutdown():
+
+            '''
+            Publish upcoming red lights at camera frequency.
+            Each predicted state has to occur `STATE_COUNT_THRESHOLD` number
+            of times till we start using it. Otherwise the previous stable state is
+            used.
+            '''
+            if self.pose is not None and self.waypoints is not None and self.camera_image is not None:
```
Comment (soniclue) [inline ros/src/tl_detector/tl_detector.py]: nit: `if None not in (self.pose, self.waypoints, self.camera_image):`

File: ros/src/waypoint_updater/waypoint_updater.py [hunk 2]
```diff
@@ -188,7 +188,7 @@ def traffic_callback(self, data):
         """
         Traffic waypoints subscriber callback function.
         """
-        self.traffic_light_wp_idx = data
+        self.traffic_light_wp_idx = data.data
 
     def obstacle_callback(self, data):
```
Comment (soniclue) [inline ros/src/waypoint_updater/waypoint_updater.py:193]: can you check if the obstacle CB needs the same fix as well (even though we won't use it)?

File: ros/src/tl_detector/tl_detector.py [hunk 3]
```diff
@@ -50,7 +51,39 @@ def __init__(self):
         self.last_wp = -1
         self.state_count = 0
 
-        rospy.spin()
+        self.spin(SPIN_FREQUENCY)
```
Comment (soniclue) [inline ros/src/tl_detector/tl_detector.py]: can we move the spin() call to under __main__ (like in other python nodes)?

**Review by soniclue — APPROVED** [review]

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 2 above)
Comment (chriskcheung) [reply ros/src/waypoint_updater/waypoint_updater.py:193]: good catch

**Review by chriskcheung — APPROVED** [review]
Changes approved

File: ros/src/waypoint_updater/waypoint_updater.py (hunk 2 above)
Comment (anonymint) [reply ros/src/waypoint_updater/waypoint_updater.py:193]: I read through discussion and forum, obstacle is not being used at the moment, Ucadify plans to add this endpoint data in the future, i debugged it but nothing print it both sim tracks.

File: ros/src/tl_detector/tl_detector.py (hunk 1 above)
Comment (anonymint) [reply ros/src/tl_detector/tl_detector.py]: PR is comming

File: ros/src/tl_detector/tl_detector.py (hunk 3 above)
Comment (anonymint) [reply ros/src/tl_detector/tl_detector.py]: OK good to be consistent, PR is coming.

**Comment by anonymint** [comment]
Refactor code as per @yangchunluo comment, please approve or merge

**Pull request merged**